Append new pod conditions when deleting pods to indicate the reason for pod deletion

mimowo opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR appends pod conditions with a dedicated pod condition type `DisruptionTarget`.
Its `reason` field indicates the reason for pod termination:
- PreemptionByKubeScheduler (Pod preempted by kube-scheduler)
- DeletionByTaintManager (Pod deleted by taint manager due to NoExecute taint)
- EvictionByEvictionAPI (Pod evicted by Eviction API)
- DeletionByPodGC (an orphaned Pod deleted by PodGC)

<!--
#### Which issue(s) this PR fixes:
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*

Fixes #
-->
#### Which issue(s) this PR fixes:

- Tracking issue: https://github.com/kubernetes/enhancements/issues/3329

#### Special notes for your reviewer:

- This PR is accompanied by other PRs within KEP:
  - to use the conditions in order to determine the handling of pod failures: https://github.com/kubernetes/kubernetes/pull/111113
  - to clean up stale `DisruptionTarget` condition: https://github.com/kubernetes/kubernetes/pull/111475
- There is a new PR prepared to update the KEP according to design changes done during implementation:
https://github.com/kubernetes/enhancements/pull/3438.

#### Does this PR introduce a user-facing change?

Yes, it appends new pod conditions when deleting a pod.
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Introduction of the `DisruptionTarget` pod condition type. Its `reason` field indicates the reason for pod termination:
- PreemptionByKubeScheduler (Pod preempted by kube-scheduler)
- DeletionByTaintManager (Pod deleted by taint manager due to NoExecute taint)
- EvictionByEvictionAPI (Pod evicted by Eviction API)
- DeletionByPodGC (an orphaned Pod deleted by PodGC)
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: 
- [Usage]: <link>
- [Other doc]: <link>
-
-->
 ```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures
```

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -2430,6 +2430,8 @@ const (
 	PodReasonUnschedulable = "Unschedulable"
 	// ContainersReady indicates whether all containers in the pod are ready.
 	ContainersReady PodConditionType = "ContainersReady"
+	// EvictedByAPI indicates that the Pod was evicted by API
```
```suggestion
	// EvictedByAPI indicates that the Pod was evicted due to an API-initiated eviction.
```

alculquicondor commented on pkg/scheduler/util/utils.go:
```diff
@@ -119,6 +122,33 @@ func PatchPodStatus(ctx context.Context, cs kubernetes.Interface, old *v1.Pod, n
 }
 
+// PrepareAndDeletePod prepares and deletes the given <pod> from API server
+func PrepareAndDeletePod(ctx context.Context, cs kubernetes.Interface, pod *v1.Pod, podConditionType v1.PodConditionType) error {
```
Since multiple binaries will be using these functions, they should go in staging/src/k8s.io/component-helpers

alculquicondor commented on pkg/controller/util/node/controller_utils.go:
```diff
@@ -99,7 +102,14 @@ func DeletePods(ctx context.Context, kubeClient clientset.Interface, pods []*v1.
 // encounters one.
 func SetPodTerminationReason(ctx context.Context, kubeClient clientset.Interface, pod *v1.Pod, nodeName string) (*v1.Pod, error) {
-	if pod.Status.Reason == nodepkg.NodeUnreachablePodReason {
+	var updatedCondition bool = false
```
```suggestion
	updatedCondition := false
```

alculquicondor commented on pkg/controller/util/node/controller_utils.go:
```diff
@@ -99,7 +102,14 @@ func DeletePods(ctx context.Context, kubeClient clientset.Interface, pods []*v1.
+			Status: v1.ConditionTrue,
+		})
+	}
+	if !updatedCondition && pod.Status.Reason == nodepkg.NodeUnreachablePodReason {
```
we should keep the reason (someone might be relying on it, even though we don't provide guarantees about it)

alculquicondor replied:
ah true. Sounds good.

alculquicondor commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -233,7 +243,7 @@ func (gcc *PodGCController) gcOrphaned(ctx context.Context, pods []*v1.Pod, node
 		}
 		klog.V(2).InfoS("Found orphaned Pod assigned to the Node, deleting.", "pod", klog.KObj(pod), "node", pod.Spec.NodeName)
-		if err := gcc.deletePod(pod.Namespace, pod.Name); err != nil {
+		if err := gcc.deleteOrphanedPod(pod.Namespace, pod.Name); err != nil {
```
I think we should update the phase to Failed as well.

alculquicondor commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -292,6 +294,36 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+
+		podCopyUpdated := rest.DefaultUpdatedObjectInfo(pod, conditionAppender)
+
+		updateOptions := &metav1.UpdateOptions{}
+		_, _, err = r.store.Update(ctx, name, podCopyUpdated, rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc, false, updateOptions)
```
```suggestion
		_, _, err = r.store.Update(ctx, name, podCopyUpdated, rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc, false, &metav1.UpdateOptions{})
```

mimowo commented on pkg/controller/util/node/controller_utils.go:
```diff
@@ -99,7 +102,14 @@ func DeletePods(ctx context.Context, kubeClient clientset.Interface, pods []*v1.
+	var updatedCondition bool = false
+	if feature.DefaultFeatureGate.Enabled(features.JobBackoffPolicy) {
+		updatedCondition = utilpod.UpdatePodCondition(&pod.Status, &v1.PodCondition{
+			Type:   v1.EvictedFromUnhealtyNode,
```
It makes sense to use the same pod condition in these three situations:
- garbage collected from a node tainted as out of service
- garbage collected from unreachable node
- deleted from unreachable node by node lifecycle controller

I suggest "DeletedFromUnsuableNode" as the name. Unreachable might be misleading cause it might be
"reachable", but tainted out of service.

alculquicondor replied:
sgtm

alculquicondor commented on pkg/api/v1/pod/util.go:
```diff
@@ -360,6 +360,17 @@ func GetPodConditionFromList(conditions []v1.PodCondition, conditionType v1.PodC
 
+// UpdatePhaseAndAddConditionForFailedPod marks pod as failed and appends a condition indicating
+// the reason of the failure
+func UpdatePhaseAndAddConditionForFailedPod(status *v1.PodStatus, condition *v1.PodCondition) (updated bool) {
```
I would put this in `staging/src/component-helpers/apps`

alculquicondor commented on pkg/scheduler/util/utils.go:
```diff
@@ -119,6 +122,32 @@ func PatchPodStatus(ctx context.Context, cs kubernetes.Interface, old *v1.Pod, n
+}
+func patchPodStatusWithCondition(ctx context.Context, cs kubernetes.Interface, pod *v1.Pod, podConditionType v1.PodConditionType) error {
+	podStatusCopy := pod.Status.DeepCopy()
+	if podutil.UpdatePhaseAndAddConditionForFailedPod(podStatusCopy, &v1.PodCondition{
```
Preemption shouldn't set the phase to Failed. The pod didn't finish yet. It's responsibility of the kubelet to do that.

We can only add the phase when we know that the kubelet is down, for example when the pod is orphan or the node is unreacheable.

alculquicondor commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -282,7 +287,7 @@ func (gcc *PodGCController) gcUnscheduledTerminating(pods []*v1.Pod) {
 
 		klog.V(2).InfoS("Found unscheduled terminating Pod not assigned to any Node, deleting.", "pod", klog.KObj(pod))
-		if err := gcc.deletePod(pod.Namespace, pod.Name); err != nil {
+		if err := gcc.deletePod(pod.Namespace, pod.Name, v1.GarbageCollectedWhileTerminating); err != nil {
```
I don't think we should add a condition in this case. Something else already removed the Pod (because it has a deletionTimestamp) so it was not a disruption by the system.

alculquicondor commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -200,7 +205,7 @@ func (gcc *PodGCController) gcTerminated(pods []*v1.Pod) {
 		go func(namespace string, name string) {
 			defer wait.Done()
-			if err := gcc.deletePod(namespace, name); err != nil {
+			if err := gcc.deletePod(namespace, name, v1.GarbageCollectedTerminated); err != nil {
```
We don't need this condition. The pod already terminated on its own. It's not a disruption.

alculquicondor commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -168,7 +173,7 @@ func (gcc *PodGCController) gcTerminating(pods []*v1.Pod) {
 		go func(namespace string, name string) {
 			defer wait.Done()
-			if err := gcc.deletePod(namespace, name); err != nil {
+			if err := gcc.deletePod(namespace, name, v1.DeletedFromUnusableNode); err != nil {
```
It's worth adding a message in the condition: "Node tainted with node.kubernetes.io/out-of-service"

alculquicondor commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -233,7 +238,7 @@ func (gcc *PodGCController) gcOrphaned(ctx context.Context, pods []*v1.Pod, node
 		}
 		klog.V(2).InfoS("Found orphaned Pod assigned to the Node, deleting.", "pod", klog.KObj(pod), "node", pod.Spec.NodeName)
-		if err := gcc.deletePod(pod.Namespace, pod.Name); err != nil {
+		if err := gcc.deletePod(pod.Namespace, pod.Name, v1.DeletedFromUnusableNode); err != nil {
```
Condition message: "Node no longer exists"

alculquicondor commented on pkg/scheduler/util/utils.go:
```diff
@@ -119,6 +122,32 @@ func PatchPodStatus(ctx context.Context, cs kubernetes.Interface, old *v1.Pod, n
+}
+
+// AddPodCondition prepares pod for deletion - add a dedicated condition indicating the reason for deletion
+func AddPodCondition(ctx context.Context, cs kubernetes.Interface, pod *v1.Pod, podConditionType v1.PodConditionType) error {
```
Do not export the function unless necessary.

krmayankk commented on pkg/apis/core/types.go:
```diff
@@ -2430,6 +2430,14 @@ const (
+	// DeletedFromUnusableNode indicates that the Pod was deleted from an unusable node
+	DeletedFromUnusableNode PodConditionType = "DeletedFromUnusableNode"
+	// EvictedByAPI indicates that the Pod was evicted due to an API-initiated eviction.
+	EvictedByAPI PodConditionType = "EvictedByAPI"
```
@alculquicondor is there a documented process for reviewing new condition types . This seems like needing an api review

alculquicondor replied:
It will, indeed. We'll request one once it's past draft.

Beyond API review, there's no other process I'm aware of.

alculquicondor commented on test/integration/evictions/evictions_test.go:
```diff
@@ -351,21 +441,9 @@ func newPod(podName string) *v1.Pod {
-}
-
-func addPodConditionReady(pod *v1.Pod) {
+func addPodConditionReady(pod *v1.Pod, phase v1.PodPhase) {
```
The phase has nothing to do with the condition Ready, so I would put set it outside of the function, where needed.

```suggestion
func addPodConditionReady(pod *v1.Pod) {
```

mimowo replied:
done

alculquicondor commented on test/integration/scheduler/preemption/preemption_test.go:
```diff
@@ -454,6 +458,145 @@ func TestPreemption(t *testing.T) {
 }
 
+// TestPreemptionWithJobBackoffPolicyGate tests preemption with the JobBackoffPolicy feature enabled
+func TestPreemptionWithJobBackoffPolicyGate(t *testing.T) {
```
Instead of creating a new test, could you add a test case in TestPreemption where the feature is enabled?

mimowo replied:
Yes, although if we go this way, I would like to simplify the existing test:
- eliminate the scheduler from the pre-existing test
- use wait.PollImmidiately instead of custom "waitForPodCondition" loop
- use "node.kubernetes.io/exclude-disruption" instead of periodic heart beats because it greatly simplifies the code test. This is in fact suggested in one comment in the test:
```golang
	// Regularly send heartbeat event to APIServer so that the cluster doesn't enter fullyDisruption mode.
	// TODO(Huang-Wei): use "NodeDisruptionExclusion" feature to simply the below logic when it's beta.
```

alculquicondor replied:
> eliminate the scheduler from the pre-existing test

this is a scheduler test, we need the scheduler :)

The rest sounds good to me. Could you do it in another PR?

mimowo replied:
Oh, when I was writing my answer I was thinking about the pre-existing test in lifecycle_test. I created an issue to simplify it (https://github.com/kubernetes/kubernetes/issues/111183). 

As for the scheduler test the mentioned issues are irrelevant so it should be easy to integrate it with the existing tests. Going to try.

mimowo replied:
merged

mimowo commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -336,9 +341,21 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
+				condition := &v1.PodCondition{Type: v1.PreemptedByScheduler, Status: v1.ConditionTrue}
+				newStatus := pod.Status.DeepCopy()
+				if apipod.UpdatePodCondition(newStatus, condition) {
+					if err := utilpod.PatchStatus(ctx, cs, victim, newStatus); err != nil {
```
done

mimowo commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -336,9 +341,21 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
-			return framework.AsStatus(err)
+		} else {
+			if feature.DefaultFeatureGate.Enabled(features.JobBackoffPolicy) {
+				condition := &v1.PodCondition{Type: v1.PreemptedByScheduler, Status: v1.ConditionTrue}
```
done

mimowo commented on pkg/scheduler/schedule_one.go:
```diff
@@ -896,11 +897,11 @@ func updatePod(ctx context.Context, client clientset.Interface, pod *v1.Pod, con
 		podStatusCopy.NominatedNodeName = nominatingInfo.NominatedNodeName
 	}
-	return util.PatchPodStatus(ctx, client, pod, podStatusCopy)
+	return utilpod.PatchStatus(ctx, client, pod, podStatusCopy)
```
ok

mimowo commented on pkg/scheduler/util/utils.go:
```diff
@@ -90,35 +88,6 @@ func MoreImportantPod(pod1, pod2 *v1.Pod) bool {
 
-// PatchPodStatus calculates the delta bytes change from <old.Status> to <newStatus>,
-// and then submit a request to API server to patch the pod changes.
-func PatchPodStatus(ctx context.Context, cs kubernetes.Interface, old *v1.Pod, newStatus *v1.PodStatus) error {
```
ok

mimowo commented on test/integration/util/util.go:
```diff
@@ -364,8 +364,7 @@ func InitTestAPIServer(t *testing.T, nsPrefix string, admission admission.Interf
 		dump := sched.Cache.Dump()
-
-		return len(dump.Nodes) == 0 && len(dump.AssumedPods) == 0, nil
+		return (dump == nil || dump.Nodes == nil || len(dump.Nodes) == 0) && len(dump.AssumedPods) == 0, nil
```
I don't know about such scenario, but I thought there might be so drafted the change, I will revert it to avoid confusion and minimize the diff as it is not needed

mimowo commented on pkg/controller/util/node/controller_utils.go:
```diff
@@ -99,11 +102,25 @@ func DeletePods(ctx context.Context, kubeClient clientset.Interface, pods []*v1.
+		updated = apipod.UpdatePodCondition(&pod.Status, &v1.PodCondition{
+			Type:   v1.DeletedFromUnusableNode,
+			Status: v1.ConditionTrue,
+			Reason: nodepkg.NodeUnreachablePodReason,
```
done

mimowo commented on test/integration/node/lifecycle_test.go:
```diff
@@ -45,6 +50,301 @@ const poll = 2 * time.Second
+
+			// Run all controllers
+			go nc.Run(testCtx.Ctx)
+			go testCtx.Scheduler.Run(testCtx.Ctx)
```
Good point, I have eliminated the scheduler and the test works. The test was based on the "TestTaintBasedEvictions" which uses Scheduler, but it might be unnecessary there too.

mimowo commented on test/integration/node/lifecycle_test.go:
```diff
@@ -45,6 +50,301 @@ const poll = 2 * time.Second
+}
+
+// TestNoTaintBasedEvictionsWithJobBackoffPolicy tests for no-taint based eviction with the JobBackoffPolicy feature
+func TestNoTaintBasedEvictionsWithJobBackoffPolicy(t *testing.T) {
```
I think this is a valid test. It tests the situation when the taint-manager is disabled. Then, there is another non-taint based eviction.

alculquicondor replied:
```suggestion
func TestUnreachableNodeEvictionsWithFinalizer(t *testing.T) {
```
?

alculquicondor replied:
or `NotReady`?

mimowo replied:
I propose TestNoTaintBasedEvictionFromNotReadyNode. The "NoTaint" seems important as the NodeLifecycleController would only perform this eviction if it is started with non-default "--enable-taint-manager=false". Also, I prefer "NotReady" over "Unreachable" as NotReady is the condition which is actually checked.

mimowo commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -298,7 +304,23 @@ func (o byCreationTimestamp) Less(i, j int) bool {
 }
 
-func (gcc *PodGCController) deletePod(ctx context.Context, namespace, name string) error {
+func (gcc *PodGCController) deletePod(ctx context.Context, namespace, name string, condition *v1.PodCondition, markFailed bool) error {
```
done

mimowo commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -163,7 +167,8 @@ func (gcc *PodGCController) gcTerminating(ctx context.Context, pods []*v1.Pod) {
 		go func(namespace string, name string) {
 			defer wait.Done()
-			if err := gcc.deletePod(ctx, namespace, name); err != nil {
+			condition := &v1.PodCondition{Type: v1.DeletedFromUnusableNode, Status: v1.ConditionTrue, Message: "Node tainted with node.kubernetes.io/out-of-service"}
```
Well. technically there are scenarios in which the node is tainted "out-of-service", but the pod is not terminating due to NoExecuteTaintManager. For example:
- the pod has toleration of the "out-of-service" taint, then NoExecuteTaintManager would not react
- kube-controller-manager was started with "--start-taint-manage=false" and the pod is terminating due to another reason

However, in both cases, as the pod is terminating already then adding the condition should be responsibility of the component that triggered terminating

alculquicondor replied:
Correct, it's not this components responsibility :)

mimowo commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -292,6 +293,41 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+			return podObj, nil
+		}
+
+		podCopyUpdated := rest.DefaultUpdatedObjectInfo(pod, conditionAppender)
```
The API of the rest.store.Update accepts an instance of rest.UpdatedObjectInfo, DefaultUpdatedObjectInfo is the default implementation of the interface.

mimowo commented on pkg/controller/util/node/controller_utils.go:
```diff
@@ -99,11 +102,25 @@ func DeletePods(ctx context.Context, kubeClient clientset.Interface, pods []*v1.
 // pod status, updates it in the apiserver, and returns an error if it
 // encounters one.
 func SetPodTerminationReason(ctx context.Context, kubeClient clientset.Interface, pod *v1.Pod, nodeName string) (*v1.Pod, error) {
```
added

alculquicondor commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -292,6 +293,39 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
 
+func addConditionAndDeletePod(r *EvictionREST, ctx context.Context, name string, validation rest.ValidateObjectFunc, options *metav1.DeleteOptions) error {
+	if feature.DefaultFeatureGate.Enabled(features.JobBackoffPolicy) {
+		if pod, err := getPod(r, ctx, name); err != nil {
+			return err
+		} else if pod != nil {
+			conditionAppender := func(_ context.Context, newObj, _ runtime.Object) (runtime.Object, error) {
```
You can reduce nesting like this:

```suggestion
		pod, err := getPod(r, ctx, name);
		if err != nil {
			return err
		} 
		conditionAppender := func(_ context.Context, newObj, _ runtime.Object) (runtime.Object, error) {
```

alculquicondor commented on pkg/scheduler/schedule_one.go:
```diff
@@ -33,14 +33,14 @@ import (
 	internalqueue "k8s.io/kubernetes/pkg/scheduler/internal/queue"
 	"k8s.io/kubernetes/pkg/scheduler/metrics"
-	"k8s.io/kubernetes/pkg/scheduler/util"
+	utilpod "k8s.io/kubernetes/pkg/util/pod"
```
revert

alculquicondor commented on test/integration/evictions/evictions_test.go:
```diff
@@ -334,6 +339,86 @@ func TestEvictionVersions(t *testing.T) {
 }
 
+// func TestEvictionWithJobBackoffPolicy tests the behaviour when the JobBackoffPolicy feature is enabled
+func TestEvictionWithJobBackoffPolicy(t *testing.T) {
```
```suggestion
func TestEvictionWithFinalizers(t *testing.T) {
```

Eventually the feature gate will go away, but we want to keep this test.

alculquicondor commented on test/integration/node/lifecycle_test.go:
```diff
@@ -45,6 +50,291 @@ const poll = 2 * time.Second
 type podCondition func(pod *v1.Pod) (bool, error)
 
+// TestTaintBasedEvictionsWithJobBackoffPolicy tests taint-based eviction with the JobBackoffPolicy feature
+func TestTaintBasedEvictionsWithJobBackoffPolicy(t *testing.T) {
```
```suggestion
func TestTaintBasedEvictionsWithFinalizers(t *testing.T) {
```

mimowo replied:
I suggest the "TestNoExecuteTaintBasedEviction" as the name. Finalizers are rather technical means to test eviction triggered by adding the "NoExecute" taint, rather than the essence of the test.

alculquicondor replied:
I just realized that this test might not require finalizers. Because there is no kubelet, the controller would simply cause the pod to get a DeletionTimestamp, but it wouldn't be actually deleted.

alculquicondor commented on test/integration/node/lifecycle_test.go:
```diff
@@ -45,6 +50,291 @@ const poll = 2 * time.Second
+			testPod := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:       "testpod",
+					Finalizers: []string{"foo.bar/some-finalizer"},
```
Use the same finalizer as the other test.

alculquicondor commented on test/integration/util/util.go:
```diff
@@ -364,7 +364,6 @@ func InitTestAPIServer(t *testing.T, nsPrefix string, admission admission.Interf
 func WaitForSchedulerCacheCleanup(sched *scheduler.Scheduler, t *testing.T) {
 	schedulerCacheIsEmpty := func() (bool, error) {
 		dump := sched.Cache.Dump()
```
revert, since it's unrelated.

mimowo commented on pkg/controller/nodelifecycle/scheduler/taint_manager.go:
```diff
@@ -120,6 +124,22 @@ func deletePodHandler(c clientset.Interface, emitEventFunc func(types.Namespaced
+	if feature.DefaultFeatureGate.Enabled(features.JobBackoffPolicy) {
+		if pod, err := getPod(name, ns); err != nil {
+			return err
+		} else if pod != nil {
```
updated

alculquicondor commented:
/retest
#111154

alculquicondor commented on pkg/controller/podgc/gc_controller_test.go:
```diff
@@ -317,11 +347,16 @@ func TestGCOrphaned(t *testing.T) {
-				nodeList.Items = append(nodeList.Items, *node)
+				nodes = append(nodes, node)
+			}
+			pods := make([]*v1.Pod, 0)
```
same as above

mimowo commented on pkg/controller/podgc/gc_controller_test.go:
```diff
@@ -136,29 +151,30 @@ func TestGCTerminated(t *testing.T) {
-				t.Errorf("[%v]pod's deleted expected and actual did not match.\n\texpected: %v\n\tactual: %v",
-					i, test.deletedPodNames.List(), deletedPodNames)
-			}
+			verifyDeletedAndPatchedPods(t, client, test.name, test.deletedPodNames, test.patchedPodNames)
```
cleaned up

mimowo commented on pkg/controller/podgc/gc_controller_test.go:
```diff
@@ -568,24 +630,63 @@ func TestGCTerminating(t *testing.T) {
+	return fake.NewSimpleClientset(nodeList, podList)
+}
+
+func getPatchedPodNames(client *fake.Clientset) (patchedPodNames []string) {
```
cleaned up

mimowo commented:
/assign @lavalamp 
Please review the API

mimowo commented:
/assign @aojea 
/assign @liggitt
Please review / approve

alculquicondor commented on test/integration/node/lifecycle_test.go:
```diff
@@ -45,6 +50,291 @@ const poll = 2 * time.Second
+			testPod := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:       "testpod",
+					Finalizers: []string{"test.k8s.io/finalizer"},
```
That makes sense. Thanks for the investigation.

mimowo replied:
I will add two comments:
- this setting pod.status.phase=Failed is hidden behind the JobBackoffPolicy feature gate for now
- this change is not essential for our KEP, as the terminating orphan pods are already accounted as failed by the job controller: https://github.com/kubernetes/kubernetes/blob/b96a04df901c6facf0fe66b3c3c877eb8651e267/pkg/controller/job/job_controller.go#L1296

alculquicondor replied:
ahhh yes, I did that more as a hack. I would like to have this be in the GC instead.

@liggitt for thoughts on whether it's a concern.

mimowo replied:
Maybe better if it is a dedicated PR/KEP which fixes it though? Then, we could also cleanup the hack in the job controller?

mimowo commented:
/assign deads2k
Please review

mimowo commented:
/retest

alculquicondor commented:
Friendly reminder:

> Please add more details in the release notes about which conditions we are adding and in which cases.

alculquicondor commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -298,7 +306,23 @@ func (o byCreationTimestamp) Less(i, j int) bool {
 }
 
-func (gcc *PodGCController) deletePod(ctx context.Context, namespace, name string) error {
+func (gcc *PodGCController) deletePodWithCondition(ctx context.Context, namespace, name string, condition *v1.PodCondition) error {
```
```suggestion
func (gcc *PodGCController) failAndDeletePodWithCondition(ctx context.Context, namespace, name string, condition *v1.PodCondition) error {
```

And probably beneficial to add a comment.

mimowo replied:
I have applied the change for now, but I think it would be better to fix this issue (and set the phase=Failed) in a dedicated PR / KEP, so that we can also simplify the job_controller logic.

alculquicondor commented on pkg/controller/util/node/controller_utils.go:
```diff
@@ -99,13 +101,29 @@ func DeletePods(ctx context.Context, kubeClient clientset.Interface, pods []*v1.
+			updated = true
+		}
+	}
+	if pod.Status.Reason != nodepkg.NodeUnreachablePodReason || pod.Status.Message != message {
```
In the past, we only checked for the reason to have changed. I think what you have is technically the same, because we are not changing the message, but just to be safe, keep it as it was.

mimowo replied:
ok, updated

alculquicondor commented on pkg/controller/util/node/controller_utils_test.go:
```diff
@@ -0,0 +1,269 @@
+	testCases := map[string]struct {
+		enableJobBackoffPolicy bool
+		initStatus             v1.PodStatus
+		expectedUpdate         bool
```
nit:
```suggestion
		wantUpdate         bool
```

alculquicondor commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -336,9 +339,25 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
+					}
+				}
+			}
+			if err := util.DeletePod(ctx, cs, victim); err != nil {
```
One concern specific to the scheduler (as opposed to the other controllers). If this Delete fails (for example, because the apiserver is suddenly unavailable, or the scheduler died), it's possible that the scheduler decides to preempt a different pod next time (or none). Then, this pod would be left with a condition that the pod was preempted, when it actually wasn't.

The solution would be to have a small controller in kube-controller-manager that either completes a preemption (does a delete) or removes the condition (or set it to false) if the condition timestamp is old enough and there is no deletion timestamp.

I would not block on this for alpha, but we probably need it for beta graduation.

mimowo replied:
Makes sense, I would vote for updating the status of the condition as false. 
For now I have added a note about this concern and updated the graduation criteria in the KEP update: https://github.com/kubernetes/enhancements/pull/3438

mimowo replied:
This is being covered in this PR: https://github.com/kubernetes/kubernetes/pull/111475

alculquicondor commented on test/integration/node/lifecycle_test.go:
```diff
@@ -44,6 +49,290 @@ const poll = 2 * time.Second
+				t.Errorf("Failed to taint node in test %s <%s>, err: %v", name, nodes[nodeIndex].Name, err)
+			}
+
+			err = wait.PollImmediate(time.Second, time.Second*20, func() (bool, error) {
```
The wait should be for the pod to have a deletion timestamp.

After that, outside of the wait, you can check whether the pod has or doesn't have the condition.

Otherwise, your check for `!test.enableJobBackoffPolicy && cond == nil` would be true even if the controller does nothing.

mimowo replied:
indeed, fixed

alculquicondor commented on test/integration/node/lifecycle_test.go:
```diff
@@ -44,6 +49,290 @@ const poll = 2 * time.Second
+				if e != nil {
+					return false, e
+				}
+				if updatedPod == nil {
+					return true, fmt.Errorf("Pod '%v' is evicted", testPod.Name)
+				}
```
This wouldn't happen, because there was no error.

mimowo replied:
fixed

alculquicondor commented on test/integration/scheduler/preemption/preemption_test.go:
```diff
@@ -430,7 +461,22 @@ func TestPreemption(t *testing.T) {
 			// Wait for preemption of pods and make sure the other ones are not preempted.
 			for i, p := range pods {
 				if _, found := test.preemptedPodIndexes[i]; found {
-					if err = wait.Poll(time.Second, wait.ForeverTestTimeout, podIsGettingEvicted(cs, p.Namespace, p.Name)); err != nil {
```
Should we change `podIsGettingEvicted` instead?

mimowo replied:
no need to, reverted the change and inspect the pod condition outside the poll as in other integration tests now

mimowo commented on pkg/controller/nodelifecycle/scheduler/taint_manager_test.go:
```diff
@@ -765,7 +779,7 @@ func TestEventualConsistency(t *testing.T) {
 			oldNode:      testutil.NewNode("node1"),
 			newNode:      addTaintsToNode(testutil.NewNode("node1"), "testTaint1", "taint1", []int{1}),
-			expectDelete: false,
+			expectDelete: true,
```
Yes, it is not checked currently on master: https://github.com/kubernetes/kubernetes/blob/303f47c0c069c80ec6d2b3879bc879924d555a8a/pkg/controller/nodelifecycle/scheduler/taint_manager_test.go#L818

mimowo commented on pkg/controller/nodelifecycle/scheduler/taint_manager_test.go:
```diff
@@ -826,15 +832,26 @@ func TestEventualConsistency(t *testing.T) {
-			if action.GetVerb() == "delete" && action.GetResource().Resource == "pods" {
-				podDeleted = true
-			}
+		close(stopCh)
```
Sure, ticketed: https://github.com/kubernetes/kubernetes/issues/111308

mimowo commented on pkg/controller/nodelifecycle/scheduler/taint_manager_test.go:
```diff
@@ -826,15 +832,26 @@ func TestEventualConsistency(t *testing.T) {
-		close(stopCh)
+	}
+	if podPatched != expectPatch {
+		t.Errorf("[%v]Unexpected test result. Expected patch %v, got %v", description, expectPatch, podPatched)
```
Ticketed, same issue as above

mimowo commented:
/retest

mimowo commented:
/retest

mimowo commented:
/retest

alculquicondor commented:
/label api-review

mimowo commented:
/retest

mimowo commented on pkg/apis/core/types.go:
```diff
@@ -2430,6 +2430,14 @@ const (
 	PodReasonUnschedulable = "Unschedulable"
 	// ContainersReady indicates whether all containers in the pod are ready.
 	ContainersReady PodConditionType = "ContainersReady"
+	// PreemptedByScheduler indicates that the Pod was preempted
```
The condition types are removed from types.go for now.

mimowo commented on pkg/apis/core/types.go:
```diff
@@ -2430,6 +2430,14 @@ const (
+	// EvictedByAPI indicates that the Pod was evicted due to an API-initiated eviction.
+	EvictedByAPI PodConditionType = "EvictedByAPI"
 )
 
```
We introduce a single "TargettedForDisruption" single condition type and the more granual reason of pod termination
is kept in the reason field. Currently supported reasons:
- PreemptedByScheduler (Pod preempted by kube-scheduler)
- DeletedByTaintManager (Pod evicted by kube-controller-manager due to taints)
- EvictedByEvictionAPI (Pod deleted by Eviction API)
- DeletedByGarbageCollector (an orphaned Pod deleted by GC)

mimowo commented on pkg/apis/core/types.go:
```diff
@@ -2430,6 +2430,14 @@ const (
+	// EvictedByTaints indicates that the Pod was evicted due to taints
+	EvictedByTaints PodConditionType = "EvictedByTaints"
+	// DeletedFromUnusableNode indicates that the Pod was deleted from an unusable node
+	DeletedFromUnusableNode PodConditionType = "DeletedFromUnusableNode"
```
The code modifications of node_lifecyle_controller when the taint manager is disabled are reverted. We will make it
impossible to disable taint manager in a dedicated PR.

mimowo replied:
This is addressed by this PR: https://github.com/kubernetes/kubernetes/pull/111411

mimowo commented on pkg/apis/core/types.go:
```diff
@@ -2430,6 +2430,14 @@ const (
+	// PreemptedByScheduler indicates that the Pod was preempted
+	PreemptedByScheduler PodConditionType = "PreemptedByScheduler"
+	// EvictedByTaints indicates that the Pod was evicted due to taints
+	EvictedByTaints PodConditionType = "EvictedByTaints"
```
Makes sense, currently we have a single pod condition type `TargettedForDisruption` for the ease of use.
For more granular reasons for pod termination we use the `reason` field, with the following options:
- PreemptedByScheduler (Pod preempted by kube-scheduler)
- DeletedByTaintManager (Pod evicted by kube-controller-manager due to taints)
- EvictedByEvictionAPI (Pod deleted by Eviction API)
- DeletedByGarbageCollector (an orphaned Pod deleted by GC)

mimowo commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -292,6 +293,39 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
 	return &metav1.Status{Status: metav1.StatusSuccess}, nil
 }
 
+func addConditionAndDeletePod(r *EvictionREST, ctx context.Context, name string, validation rest.ValidateObjectFunc, options *metav1.DeleteOptions) error {
```
We are going to work on this in a dedicated PR for this KEP. Do you think it needs to be in Alpha or Beta is also fine?

alculquicondor replied:
Yes, it should be ready for Alpha, but we can do it in a separate PR.

mimowo replied:
It is addressed in this PR: https://github.com/kubernetes/kubernetes/pull/111475

alculquicondor commented on pkg/controller/nodelifecycle/scheduler/taint_manager.go:
```diff
@@ -115,6 +119,27 @@ func deletePodHandler(c clientset.Interface, emitEventFunc func(types.Namespaced
+		}
+		newStatus := pod.Status.DeepCopy()
+		if apipod.UpdatePodCondition(newStatus, &v1.PodCondition{
+			Type:    "TargettedForDisruption",
```
even if we are not exporting the condition, you can still have it as a constant on each of the packages that need it, so you can reuse it for tests.

And it's `Targeted`

mimowo replied:
updated

alculquicondor commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -173,7 +175,7 @@ func (gcc *PodGCController) gcTerminating(ctx context.Context, pods []*v1.Pod) {
 		go func(namespace string, name string) {
 			defer wait.Done()
-			if err := gcc.deletePod(ctx, namespace, name); err != nil {
+			if err := gcc.failAndDeletePodWithCondition(ctx, namespace, name, nil); err != nil {
```
worth adding a `failAndDeletePod` that doesn't require a condition and calls the other function.

mimowo replied:
updated

alculquicondor commented on pkg/controller/nodelifecycle/scheduler/taint_manager.go:
```diff
@@ -115,6 +119,27 @@ func deletePodHandler(c clientset.Interface, emitEventFunc func(types.Namespaced
+		if apipod.UpdatePodCondition(newStatus, &v1.PodCondition{
+			Type:    "TargettedForDisruption",
+			Status:  v1.ConditionTrue,
+			Reason:  "DeletedByTaintManager",
```
Maybe just `ByTaintManager` or `DeletionByTaintManager`, because the actual deletion is coming later.

The same can be said for the message.

mimowo replied:
I prefer "DeletionByTaintManager" over "ByTaintManager", as it is a proper reason description without the context. It also feels more consistent naming with the reasons introduced by Kubelet (which include the component name and a short reason description, like KubeletHasInsufficientMemory): https://github.com/kubernetes/kubernetes/blob/c633f7124d9125067d89bd1e92cdff882144ced6/pkg/kubelet/nodestatus/setters.go#L594

alculquicondor commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -238,7 +240,13 @@ func (gcc *PodGCController) gcOrphaned(ctx context.Context, pods []*v1.Pod, node
 		}
 		klog.V(2).InfoS("Found orphaned Pod assigned to the Node, deleting.", "pod", klog.KObj(pod), "node", pod.Spec.NodeName)
-		if err := gcc.deletePod(ctx, pod.Namespace, pod.Name); err != nil {
+		condition := &v1.PodCondition{
```
@liggitt what do you think about optimizing here and only adding the condition if the pod has any finalizer?

liggitt replied:
optimizations are good if they make sense. I don't know enough about the observers of this condition to know if they would react differently to `orphaned pod → added condition → delete-with-0-grace-period` than a simple `orphaned pod → delete-with-0-grace-period` sequence.

mimowo replied:
Seems it requires some investigation to understand the balance between optimization and code complication to avoid a premature optimization. Maybe investigating the optimization should be another graduation crateria for Beta? @alculquicondor  @liggitt ?

alculquicondor replied:
An observer could benefit from that extra "added condition" notification, although it's not guaranteed (their controller could be down, for example). In any case, it makes sense to keep it in all cases to stay consistent.

lavalamp replied:
Users/controllers can't depend on a separate notification, since the (client-side) watch infrastructure may combine changes, or they may get disconnected and need to relist.

alculquicondor commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -308,7 +316,26 @@ func (o byCreationTimestamp) Less(i, j int) bool {
+		// Mark the pod as failed - this is especially important in case the pod
+		// is orphaned, in which case the pod would remain in the Running phase
+		// forever as there is no kubelet running to change the phase.
+		if pod.Status.Phase != v1.PodSucceeded {
```
@smarterclayton you might be interested in this part.

In general, this causes the apiserver to do an immediate deletion of the object. But GC always did that by setting the grace period to zero.

alculquicondor commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -336,9 +339,26 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
+		} else {
+			if feature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) {
+				condition := &v1.PodCondition{
+					Type:    "TargettedForDisruption",
```
```suggestion
					Type:    "TargetedForDisruption",
```

mimowo replied:
updated

alculquicondor commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -336,9 +339,26 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
+				condition := &v1.PodCondition{
+					Type:    "TargettedForDisruption",
+					Status:  v1.ConditionTrue,
+					Reason:  "PreemptedByScheduler",
```
`ByKubeScheduler` or `PreemptionByKubeScheduler`

mimowo replied:
updated

alculquicondor commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -336,9 +339,26 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
+					Type:    "TargettedForDisruption",
+					Status:  v1.ConditionTrue,
+					Reason:  "PreemptedByScheduler",
+					Message: "Preempted by scheduler to accommodate a Pod with higher priority",
```
```suggestion
					Message: "To be preempted by scheduler to accommodate a Pod with higher priority",
```

mimowo replied:
updated

alculquicondor commented on pkg/util/pod/pod.go:
```diff
@@ -66,5 +66,6 @@ func preparePatchBytesForPodStatus(namespace, name string, uid types.UID, oldPod
 		return nil, false, fmt.Errorf("failed to CreateTwoWayMergePatch for pod %q/%q: %v", namespace, name, err)
 	}
-	return patchBytes, bytes.Equal(patchBytes, []byte(fmt.Sprintf(`{"metadata":{"uid":%q}}`, uid))), nil
+	unchanged := bytes.Equal(patchBytes, []byte("{}")) || bytes.Equal(patchBytes, []byte(fmt.Sprintf(`{"metadata":{"uid":%q}}`, uid)))
```
🤔 was this change necessary?

mimowo replied:
This change is not necessary and I actually prefer to avoid making the change without protecting it by the feature gate. 

When I was adding the unit tests in gc_controller_test.go I realized that the patch requests are sent if nothing changes. 

In order to avoid the unnecessary patch I now check the result of UpdatePodCondition in gc_controller.go. 

Still, in the long run it seems sensible to make the low-level check to avoid complications of the callers code.

alculquicondor commented on test/integration/node/lifecycle_test.go:
```diff
@@ -44,6 +48,144 @@ const poll = 2 * time.Second
 type podCondition func(pod *v1.Pod) (bool, error)
 
+// TestNoExecuteTaintBasedEviction tests taint-based eviction for a node tainted NoExecute
+func TestNoExecuteTaintBasedEviction(t *testing.T) {
```
Could this have been added as a test case for `TestTaintBasedEvictions`?

mimowo replied:
We could, but it would risk making the test flaky due to timeout. The current TestTaintBasedEvictions takes already on my machine 41s which is above the default timeout of 30s.

Also, the test TestTaintBasedEvictions is ticketed for simplifications here: [Simplify the "TestTaintBasedEvictions" integration test in lifecycle_test.go](https://github.com/kubernetes/kubernetes/issues/111183). I think the new test I introduce could serve as an example of how to refactor the existing one. Once the test is refactored it could take significantly less time (due to elimination of kube scheduler and periodic heartbeats) and then merging the tests could be feasible.

mimowo replied:
When makes the new test also sufficiently different from the existing one is that the test simulates adding a custom NoExecute taint by the user, whereas the existing test awaits (https://github.com/kubernetes/kubernetes/blob/7156c96e5d8f4fe58a36000ff53af7e21743f49b/test/integration/node/lifecycle_test.go#L273) for the taints to be added by the node_lifecycle_controller (https://github.com/kubernetes/kubernetes/blob/7156c96e5d8f4fe58a36000ff53af7e21743f49b/pkg/controller/nodelifecycle/node_lifecycle_controller.go#L708)

alculquicondor commented on test/integration/podgc/podgc_test.go:
```diff
@@ -36,144 +37,181 @@ import (
+		enableJobPodFailurePolicy bool
+	}{
+		"JobPodFailurePolicy enabled": {
+			enableJobPodFailurePolicy: true,
```
Shouldn't we test that the phase is set to Failed? And then we would need a finalizer.

mimowo replied:
The test is extended for testing if the phase is changed with the finalizer. Still, I keep the test for the scenario when there is no finalizer to test that the pod gets deleted quickly (grace period is 0). This way I don't drop this scenario which was covered before this PR.

mimowo commented:
/retest

mimowo commented:
/remove api-review
/remove kind/api-change

alculquicondor commented:
/remove-label api-review
/remove-kind api-change

alculquicondor commented on test/integration/evictions/evictions_test.go:
```diff
@@ -47,15 +48,28 @@ import (
+	// The constant is not exported in order to prevent its external use until
+	// the name is accepted within the KEP-3329. Once the name is accepted the
+	// declaration will be moved to the core API.
+	targetedForDisruption api.PodConditionType = "TargetedForDisruption"
```
```suggestion
	targetedForDisruption v1.PodConditionType = "TargetedForDisruption"
```

alculquicondor commented on test/integration/podgc/podgc_test.go:
```diff
@@ -28,152 +28,238 @@ import (
+				t.Fatalf("Pod %v/%v does has the unexpected condition: %v", pod.Namespace, pod.Name, targetedForDisruption)
+			}
+			if test.wantPhase != pod.Status.Phase {
+				t.Fatalf("Enexpected phase for pod '%v'. Expected: '%v;, got: '%v'", klog.KObj(pod), test.wantPhase, pod.Status.Phase)
```
Prefer `Errorf` when checking the expectations of the test. This allows the test to continue and get the remaining errors.

`Fatalf` makes sense when the test setup fails, because there is no point continuing.

```suggestion
				t.Errorf("Unexpected phase for pod %q. Got %q, want %q'", klog.KObj(pod), pod.Status.Phase, test.wantPhase)
```

alculquicondor commented on test/integration/podgc/podgc_test.go:
```diff
@@ -28,152 +28,238 @@ import (
+					}
+					return test.wantPhase == pod.Status.Phase, nil
+				})
+				if err != nil {
```
Check the condition too.

mimowo replied:
I've added a check that the condition is not there. Currently, we append the condition in gc_controller only for orphaned pods, as it seems to be the responsibility of other components to add an appropriate condition in such a case. Maybe, actually, we should also set the phase to Failed to orphaned pods only @alculquicondor ?

alculquicondor replied:
Oops, right, there is no condition here.

I think we should set the phase to Failed here, because the pod is being fully removed regardless.

liggitt commented on pkg/features/kube_features.go:
```diff
@@ -801,6 +801,14 @@ const (
+	//
+	// Allow users to specify handling of pod failures based on container exit codes
+	// and pod end state.
+	JobPodFailurePolicy featuregate.Feature = "JobPodFailurePolicy"
```
since this feature is primarily controlling whether pod-controllers add conditions, I'd suggest a less job-centric name

liggitt replied:
I'll defer to @alculquicondor on the feature name, but something more pod-centric like that seems better to me

alculquicondor replied:
Should we consider two feature gates?

`PodDisruptionConditions` and `JobPodFailurePolicy`?

If we go with one, I would go with `PodDisruptionConditions`

lavalamp replied:
Yes, I think you need two feature gates.

mimowo replied:
Makes sense, I have added `PodDisruptionConditions` gate for this PR (`JobPodFailurePolicy` remains to control the job_controller PR). Also, the KEP PR is updated accordingly: https://github.com/kubernetes/enhancements/pull/3438

mimowo replied:
The introduction of the second feature gate introduced a merge conflict between the two PRs. Should we consider declaration of the two feature gates on both PRs so that they merge without conflicts once lgtmed/approved?

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -280,6 +280,9 @@ func buildControllerRoles() ([]rbacv1.ClusterRole, []rbacv1.ClusterRoleBinding)
 			rbacv1helpers.NewRule("list", "watch", "delete").Groups(legacyGroup).Resources("pods").RuleOrDie(),
 			rbacv1helpers.NewRule("get", "list").Groups(legacyGroup).Resources("nodes").RuleOrDie(),
+			// used for pod deletion
+			rbacv1helpers.NewRule("patch").Groups(legacyGroup).Resources("pods/status").RuleOrDie(),
```
pod GC doesn't have permission to write to pods at all today... only add this permission if the pod condition feature is enabled

mimowo replied:
Done, it is now protected by the feature gate flag? Should we also make an analogous change to only add "patch" to node-controller?

liggitt replied:
> Should we also make an analogous change to only add "patch" to node-controller?

That's less of a big deal... it already had update permission, adding `patch` is not granting it more power, just an additional way to accomplish the same thing.

This line is adding write permission where it didn't exist previously, so gating makes more sense

lavalamp commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -50,6 +52,15 @@ const (
+	// The constant is not exported in order to prevent its external use until
+	// the name is accepted within the KEP-3329. Once the name is accepted the
+	// declaration will be moved to the core API.
+	targetedForDisruption v1.PodConditionType = "TargetedForDisruption"
```
I saw the old comment about not putting this in a central file until it's beta but I'm not convinced declaring it in N places is better.

The primary problem about exposing it is that then people might use it and it would be hard to change the name. But I don't think multiple declarations solves that problem at all.

mimowo replied:
My thinking is that it solves the problem of using the pod condition directly from code at compile time without any notification. Still, one could use the pod condition at runtime, but needs to enable the feature gate flag in order to trigger appending of these pod conditions. Multiple declaration places could be hard to maintain in the long run, but the idea is that we eliminate the duplication in Beta. Any other ideas?

mimowo replied:
As for the name, I like `DisruptionTarget`, @alculquicondor ?

lavalamp replied:
> solves the problem of using the pod condition directly from code at compile time without any notification

You could make the name of the constant AlphaWillBeRenamedNoCompatGuaranteeDisruptionTargetCondition or something like that.

mimowo replied:
Indeed, it avoids the duplication and  will prevent users for committing to the name, but it may look a little bit awkward in the core API.

I'm personally fine with both approaches:
1. DisruptionTarget duplicated is each package for Alpha
2. using the proposed name, but I'm thinking about `DisruptionTargetToBeRenamedForBeta` it seems clear enough that if it is to be renamed, then no compat guarantee

Any further opinions?

lavalamp replied:
Personally I prefer option 2, @liggitt may have a different opinion though

alculquicondor replied:
Re name: The guideline is the following (from https://github.com/kubernetes/community/blob/308bb1109fe07cf3ecf14381129ecabac98de425/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties):

> Condition type names should describe the current observed state of the resource, rather than describing the current state transitions. This typically means that the name should be an adjective ("Ready", "OutOfDisk") or a past-tense verb ("Succeeded", "Failed") rather than a present-tense verb ("Deploying"). 

I think both qualify, then favoring the shorter `DisruptionTarget` should be fine.

I actually like `AlphaWillBeRenamedNoCompatGuaranteeDisruptionTargetCondition`... very in-your-face. But maybe something like `AlphaNoCompatGuaranteeDisruptionTargetCondition` and a comment should suffice. @liggitt are you ok with this or would you rather have the variables private?

mimowo replied:
For now I have updated the private pod condition name from `TargetedForDisruption` to `DisruptionTarget`. I will wait for the @liggitt opinion and general consensus before introducing the in-face name into the core API. I'm ok with `AlphaNoCompatGuaranteeDisruptionTargetCondition` as the name, but I think the last `Condition` word is redundant, cause this is conveyed by the type already.

liggitt replied:
Knowing how our API packages get consumed, I really didn't want to export things in from core/v1 that aren't stable (either stable go field names that represent unstable functionality, or unstable names we intend to remove/break callers of in the future). That's not an official API guidelines position, just my $0.02 having heard friction reports of people using our libraries.

If other API folks think explicitly unstable constant names are reasonable, I won't block on that. I'd probably recommend a simple `Alpha...` prefix to the go field name and explanatory godoc sufficient in that case.

lavalamp replied:
We intend to break go compatibility rules, and not the k8s API rules. That's why I suggested an especially unwieldy name. I went a bit overboard, but I'd use more than just "Alpha" to warn people.

I don't have a super strong opinion, I just wouldn't say that copy-paste inheritance is clearly better than breaking compatibility rules.

mimowo replied:
I have pushed a commit which introduces the new pod condition type into the core APIs (I make use of the new constants in the next commit):
https://github.com/kubernetes/kubernetes/pull/110959/commits/c96585e316f3f83db98462ff4876ff5235fede3f

A couple of notes:
1. I've reviewed how other pod condition types were introduced. For reference:
- ContainersReady (2018)
  - Commit: https://github.com/kubernetes/kubernetes/commit/6b08ef575fe52f49acac53583d6b954ad27a8806
  - PR: https://github.com/kubernetes/kubernetes/pull/64646
  - Approach: Direct introduction into into core API
- Unschedulable (2016)
  - Commit: https://github.com/kubernetes/kubernetes/commit/859077b70183993f492fd2d3de66b9effcd6fef1
  - PR:https://github.com/kubernetes/kubernetes/pull/34634
  - Approach: Replaced a string
- Initialized (2016)
  - Commit: https://github.com/kubernetes/kubernetes/commit/6685715c4ca1f143108273edd038806b2bb5fd72
  - PR: https://github.com/kubernetes/kubernetes/pull/23567
  - Approach: Direct introduction into core API
- PodScheduled (2016)
  - Commit: https://github.com/kubernetes/kubernetes/commit/a80b1798c45f88e6b38261a1657c0dfa787de136
  - PR: https://github.com/kubernetes/kubernetes/pull/24459
  - Approach: Direct introduction into core API

2. I shortened the name and removed the `Condition` word from the end as the other pod condition types don't have it

3. I keep the string value represented by the constant as `DisruptionTarget`, but can update. Any opinions?

lavalamp commented on pkg/controller/nodelifecycle/scheduler/taint_manager.go:
```diff
@@ -115,6 +128,27 @@ func deletePodHandler(c clientset.Interface, emitEventFunc func(types.Namespaced
+			Type:    targetedForDisruption,
+			Status:  v1.ConditionTrue,
+			Reason:  "DeletionByTaintManager",
+			Message: "To be deleted by taint manager due to the NoExecute taint",
```
"taint manager: deleting due to NoExecute taint"

mimowo replied:
updated, just capitalized the first letter

lavalamp commented on pkg/controller/nodelifecycle/scheduler/taint_manager.go:
```diff
@@ -115,6 +128,27 @@ func deletePodHandler(c clientset.Interface, emitEventFunc func(types.Namespaced
+			Reason:  "DeletionByTaintManager",
+			Message: "To be deleted by taint manager due to the NoExecute taint",
+		}) {
+			if _, _, _, err = utilpod.PatchPodStatus(ctx, c, pod.Namespace, pod.Name, pod.UID, pod.Status, *newStatus); err != nil {
```
I would prefer to see this code written with SSA (with its client), I think the code would be shorter and have less chance of conflicts, and you can omit the GET. Did you try that?

alculquicondor replied:
SSA: https://kubernetes.io/docs/reference/using-api/server-side-apply/

mimowo replied:
Haven't tried this yet - I was inspired by the code used to add other pod conditions. Could this be considered as an implementation improvement attempted for Beta? I'm thinking about it due to the time constraint, but would like to try.

alculquicondor replied:
not sure how relevant this is, but if we were to change it in the next release, the permissions for kcm would have to change a little. In any case, I'm leaving this for @lavalamp to decide if it can be done for beta.

lavalamp replied:
Yes, this can wait for beta. I didn't comment on all the other locations, at least some of them could probably also benefit, but at some point it starts being a lot of work.

mimowo replied:
Ok, I've added a it as a graduation criteria for Beta in the KEP update PR: https://github.com/kubernetes/enhancements/pull/3438:
```
Review and implement if feasible adding of pod conditions with the use of [SSA](https://kubernetes.io/docs/reference/using-api/server-side-apply/) client.
```

krmayankk replied:
Is there an example of using SSA for get-update in the existing code ?

lavalamp replied:
SSA means you can omit the GET.

https://kubernetes.io/blog/2021/08/06/server-side-apply-ga/#server-side-apply-support-in-client-go

lavalamp commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -238,7 +249,13 @@ func (gcc *PodGCController) gcOrphaned(ctx context.Context, pods []*v1.Pod, node
+			Type:    targetedForDisruption,
+			Status:  v1.ConditionTrue,
+			Reason:  "DeletionByPodGC",
+			Message: "To be deleted by pod garbage-collector as the node no longer exists",
```
"podgc: node no longer exists"

Note that this is not the regular garbage collector and the current message will be confusing.

mimowo replied:
indeed, updated, just capitalized the first letter

lavalamp commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -292,6 +302,41 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
+				Type:    targetedForDisruption,
+				Status:  api.ConditionTrue,
+				Reason:  "EvictionByEvictionAPI",
+				Message: "To be evicted by eviction API",
```
"Eviction API: evicting"

lavalamp commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -336,9 +348,26 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
+					Type:    targetedForDisruption,
+					Status:  v1.ConditionTrue,
+					Reason:  "PreemptionByKubeScheduler",
+					Message: "To be preempted by kube-scheduler to accommodate a Pod with higher priority",
```
"kube-scheduler: preempted"

alculquicondor replied:
preempting?

mimowo replied:
Updated as "Kube-scheduler: preempting"

lavalamp replied:
Personally I'd use "kube-scheduler" to match the name of the binary, but it doesn't matter that much

mimowo commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -280,6 +280,9 @@ func buildControllerRoles() ([]rbacv1.ClusterRole, []rbacv1.ClusterRoleBinding)
 			rbacv1helpers.NewRule("get", "list").Groups(legacyGroup).Resources("nodes").RuleOrDie(),
+			// used for pod deletion
+			rbacv1helpers.NewRule("patch").Groups(legacyGroup).Resources("pods/status").RuleOrDie(),
+			rbacv1helpers.NewRule("get").Groups(legacyGroup).Resources("pods").RuleOrDie(),
```
It was needed to execute this line to get pod before patching
```golang
pod, err := gcc.kubeClient.CoreV1().Pods(namespace).Get(ctx, name, metav1.GetOptions{})
```
however, this line is not needed as gc_controller iterates over all pods and have the status. I have updated the code to not require this line and removed the rule.

liggitt commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -241,17 +241,23 @@ func buildControllerRoles() ([]rbacv1.ClusterRole, []rbacv1.ClusterRoleBinding)
+				eventsRule(),
+			},
+		}
+		if utilfeature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) {
```
since this controller already had `update` `pods/status` permission, adding `patch` to that ungated isn't a big deal

mimowo replied:
Ok, if you don't mind I prefer to stay with gating the patch for node-controller anyway so that the unprotected impact of the PR is as limited as possible

alculquicondor replied:
Actually, our side controller that knows how to remove stale conditions needs this permission even if the feature is disabled. In fact, if the feature is disabled it's more important to be able to remove the condition. So I would avoid querying the feature gate here.

alculquicondor commented on pkg/features/kube_features.go:
```diff
@@ -807,9 +807,9 @@ const (
-	// and pod end state.
-	JobPodFailurePolicy featuregate.Feature = "JobPodFailurePolicy"
+	// Enables support for appending a dedicated pod condition indicating the
+	// the pod is being deleted due to an external disruption.
```
It's not external. I would just say "disruption"

alculquicondor reviewed (commented):
Other than the remark about SSA, this LGTM

alculquicondor commented on pkg/features/kube_features.go:
```diff
@@ -802,6 +802,14 @@ const (
+	// kep: http://kep.k8s.io/3329
+	// alpha: v1.25
+	//
+	// Enables support for appending a dedicated pod condition indicating the
```
```suggestion
	// Enables support for appending a dedicated pod condition indicating that
```

alculquicondor commented on pkg/registry/core/pod/storage/eviction.go:
```diff
@@ -50,6 +52,15 @@ const (
+	// The constant is not exported in order to prevent its external use until
+	// the name is accepted within the KEP-3329. Once the name is accepted the
+	// declaration will be moved to the core API.
+	targetedForDisruption api.PodConditionType = "TargetedForDisruption"
```
```suggestion
	targetedForDisruption v1.PodConditionType = "TargetedForDisruption"
```

alculquicondor commented:
This LGTM. The only open question is here: https://github.com/kubernetes/kubernetes/pull/110959#discussion_r929157627, which is requiring @liggitt's input

mimowo commented:
/retest

alculquicondor commented:
/lgtm
/hold
for other reviewers (and squash)

mimowo commented:
squashed

alculquicondor commented:
/lgtm

mimowo commented:
/retest

mimowo commented:
/retest

alculquicondor commented:
Two things to think about:
- What should the controllers do if the DisruptionTarget condition was already added by another controller. The options are:
  - Overwrite the condition and delete (what this PR is doing)
  - Do not overwrite the condition but continue with delete (this might be the most sensible approach).
  - Do not overwrite the condition and do not delete (assuming that the controller that added it is going to do the delete).
 - The logic in the scheduler and the taint-manager are the same, so we could consider having it in [component-helpers](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/component-helpers). This might help people writing other controllers that do disruption (such as descheduler) to just use the function. The pod-gc is a bit different (because it sets phase Failed and does deletion without grace period, but this could be an option). And the apiserver is different because it uses different objects (but maybe we can use golang's generics?). But also we can leave this refactoring for beta.

liggitt commented:
> What should the controllers do if the DisruptionTarget condition was already added by another controller. The options are:
> * Overwrite the condition and delete (what this PR is doing)

I think overwriting a status other than True seems reasonable

> * Do not overwrite the condition but continue with delete (this might be the most sensible approach).

I think a pre-existing condition with type=DisruptionTarget and status=True should be left as-is, and our components should continue with deletion

liggitt commented:
API addition lgtm. Reassign me once the implementation has lgtm / approve and I can tag for the API package.

lavalamp commented:
> I think a pre-existing condition with type=DisruptionTarget and status=True should be left as-is, and our components should continue with deletion

This case means controllers are already fighting; I think we should fight only over the actual deletion and not over the condition.

liggitt commented:
> > I think a pre-existing condition with type=DisruptionTarget and status=True should be left as-is, and our components should continue with deletion
> 
> This case means controllers are already fighting; I think we should fight only over the actual deletion and not over the condition.

I think we're agreeing? I'd expect our controllers to not try to modify an existing True DisruptionTarget condition and just continue with their deletion logic

lavalamp commented:
> I think we're agreeing?

Yeah, I just wanted to point out that the case should be rare and therefore we don't need to spend too much effort optimizing it either way.

mimowo commented:
@lavalamp @liggitt @alculquicondor I assume we have a conclusion that the preferred behavior is: do not update the pod condition, if type=DisruptionTarget nad status=True, but continue the deletion. 

Is it ok for you to have this for Beta? For Beta we already intend changes in this area to commonize the code pod preparation for deletion and to use SSA client.

alculquicondor commented:
Thanks @mimowo.
Given that the codefreeze is on Tuesday (with Monday being a holiday in US and Canada) and given that this change is more of an optimization and on the nit-side, the proposal to park it for beta sounds reasonable to me.

/lgtm

alculquicondor commented:
/approve
for scheduler

alculquicondor commented:
@soltysh @janetkuo could you review/approve from sig-apps perspective?

mimowo commented:
Updated the PR to just rebase and relocate the feature gate flag in kube_features.go - to make it alphabetically placed and thus avoid conflicts with another PR: https://github.com/kubernetes/kubernetes/pull/111113.

alculquicondor commented:
/lgtm

mimowo commented:
/retest

soltysh commented on pkg/controller/podgc/gc_controller_test.go:
```diff
@@ -51,41 +52,32 @@ func NewFromClient(kubeClient clientset.Interface, terminatedPodThreshold int) (
+		name                          string
+		pods                          []nameToPhase
+		threshold                     int
+		deletedPodNames               sets.String
```
Nit: would be nice if the taint_manager_test.go is also checking deleted and patched pod names rather then just verifying the action type.

mimowo replied:
A similar testing approach is used in gc_controller_test.go. 

This seems like a pre-existing limitation in the test, which may require an extension of the fake kube client to interpret patches. thus I suggest an independent Issue to improve the tests.

soltysh reviewed (approved):
Minor nits which can be fixed in https://github.com/kubernetes/kubernetes/pull/111113 or any other followup
/lgtm
/approve
from sig-apps pov

mimowo commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -308,7 +316,30 @@ func (o byCreationTimestamp) Less(i, j int) bool {
-func (gcc *PodGCController) deletePod(ctx context.Context, namespace, name string) error {
-	klog.InfoS("PodGC is force deleting Pod", "pod", klog.KRef(namespace, name))
-	return gcc.kubeClient.CoreV1().Pods(namespace).Delete(ctx, name, *metav1.NewDeleteOptions(0))
+func (gcc *PodGCController) failAndDeletePod(ctx context.Context, pod *v1.Pod) error {
```
pushed update with suggested names

aojea commented on pkg/controller/nodelifecycle/scheduler/taint_manager.go:
```diff
@@ -115,6 +119,27 @@ func deletePodHandler(c clientset.Interface, emitEventFunc func(types.Namespaced
+func addConditionAndDeletePod(ctx context.Context, c clientset.Interface, name, ns string) (err error) {
+	if feature.DefaultFeatureGate.Enabled(features.PodDisruptionConditions) {
+		pod, err := c.CoreV1().Pods(ns).Get(ctx, name, metav1.GetOptions{})
+		if err != nil {
```
you return error if is a "NotFound" errror because the pod was already deleted, or that can not happen?

mimowo replied:
It may happen, if the pods is deleted since queuing for eviction by the taint manager: https://github.com/kubernetes/kubernetes/blob/9f9602ef830d55e844ad500c0a3ab342dd7cc98d/pkg/controller/nodelifecycle/scheduler/taint_manager.go#L382
however, the already called Delete method would also return NotFound error so it seems consistent to also return NotFound from the get.

krmayankk replied:
+1 if the error is NotFound error, we should return success and bail out

krmayankk replied:
ignore that, the default behavior is good.

alculquicondor reviewed (commented):
/lgtm

aojea commented on pkg/controller/nodelifecycle/scheduler/taint_manager_test.go:
```diff
@@ -180,8 +195,9 @@ func TestCreatePod(t *testing.T) {
 	}
 
 	for _, item := range testCases {
+		defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.PodDisruptionConditions, item.enablePodDisruptionConditions)()
```
you have to run this as a subtest , or this defer only will be executed after the whole loop , not after each test case
https://go.dev/blog/subtests

mimowo replied:
Adjusted

aojea commented on pkg/controller/nodelifecycle/scheduler/taint_manager_test.go:
```diff
@@ -269,8 +291,9 @@ func TestUpdatePod(t *testing.T) {
 	}
 
 	for _, item := range testCases {
+		defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.PodDisruptionConditions, item.enablePodDisruptionConditions)()
```
same here https://github.com/kubernetes/kubernetes/pull/110959/files#r934826728, but seems these tests were already creating multiple controllers in parallel, unless I'm getting it wrong

mimowo replied:
Adjusted

aojea commented on pkg/controller/nodelifecycle/scheduler/taint_manager_test.go:
```diff
@@ -458,6 +479,7 @@ func TestUpdateNode(t *testing.T) {
 	}
 
 	for _, item := range testCases {
+		defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.PodDisruptionConditions, item.enablePodDisruptionConditions)()
```
ditto

mimowo replied:
Adjusted

aojea commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -238,7 +240,13 @@ func (gcc *PodGCController) gcOrphaned(ctx context.Context, pods []*v1.Pod, node
+			Type:    v1.AlphaNoCompatGuaranteeDisruptionTarget,
+			Status:  v1.ConditionTrue,
+			Reason:  "DeletionByPodGC",
+			Message: "Podgc: node no longer exists",
```
I see `PodGC` in most of the logging messages, but I don't know which format is correct

mimowo replied:
Indeed, PodGC is produced in this log message: https://github.com/kubernetes/kubernetes/blob/8e14585d8700d9cf7861be66943f023d11d41075/pkg/controller/podgc/gc_controller.go#L312

Let me stick to it in the message here as well.

aojea commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -308,7 +316,30 @@ func (o byCreationTimestamp) Less(i, j int) bool {
+		}
+		if updated {
+			if _, _, _, err := utilpod.PatchPodStatus(ctx, gcc.kubeClient, pod.Namespace, pod.Name, pod.UID, pod.Status, *newStatus); err != nil {
+				return err
```
should you try to delete the pod always?

alculquicondor replied:
No, we need the information, we should let the controller retry.

mimowo replied:
Good point, @alculquicondor  @soltysh ? I suggest to maybe re-evaluate the decision for Beta. Anyway for Beta we are planning to commonize the the code for appending the pod conditions.

aojea commented on pkg/controller/podgc/gc_controller_test.go:
```diff
@@ -568,24 +607,74 @@ func TestGCTerminating(t *testing.T) {
+				nodeInformer.Informer().GetStore().Add(node)
+			}
+
+			gcc.gc(context.TODO())
```
this may leak goroutines, can we create a context with cancel and defer the cancel()?

mimowo replied:
indeed, however, this is a pre-existing issue, moved line from here: https://github.com/kubernetes/kubernetes/blob/9f9602ef830d55e844ad500c0a3ab342dd7cc98d/pkg/controller/podgc/gc_controller_test.go#L582

I will create a dedicated issue / pr to cleanup the test. Similar to this one: https://github.com/kubernetes/kubernetes/issues/111308

aojea replied:
yeah, it seems you hit a few of those :upside_down_face:

aojea commented on pkg/scheduler/framework/preemption/preemption.go:
```diff
@@ -340,9 +343,26 @@ func (ev *Evaluator) prepareCandidate(ctx context.Context, c Candidate, pod *v1.
+				if apipod.UpdatePodCondition(newStatus, condition) {
+					if err := util.PatchPodStatus(ctx, cs, victim, newStatus); err != nil {
+						klog.ErrorS(err, "Preparing pod preemption", "pod", klog.KObj(victim), "preemptor", klog.KObj(pod))
+						return framework.AsStatus(err)
```
naive question since I'm not familiar with this bits, I did a similar one before,  before the pod was always deleted, but now it seems that if the patch to the pod status fail the pod is never deleted ... is that expected?

alculquicondor replied:
That is correct. The scheduler might want to delete the same pod later. But it might decide to delete a different pod or no pod at all. That's why we will clear the condition if this happens #111475

mimowo replied:
Without the patch the delete could also fail - say due to api server being temporarily down. However, indeed, with patches we risk conflicts, which could fail more often than simple deletes. @alculquicondor  @soltysh opinions?

mimowo replied:
We could proceed with deletes if the patch fails, but I suggest to re-evaluate the decision for Beta. Anyway for Beta we are planning to commonize the the code for appending the pod conditions.

liggitt commented:
/approve
for API constant and RBAC change

@soltysh has pkg/controller approval for sig-apps
@alculquicondor has scheduler approval for sig-scheduler

liggitt commented:
I left the hold since I couldn't tell if @aojea's comments on the tests were blocking

krmayankk commented on pkg/controller/podgc/gc_controller.go:
```diff
@@ -308,7 +316,30 @@ func (o byCreationTimestamp) Less(i, j int) bool {
+	return gcc.markFailedAndDeletePodWithCondition(ctx, pod, nil)
+}
+
+func (gcc *PodGCController) markFailedAndDeletePodWithCondition(ctx context.Context, pod *v1.Pod, condition *v1.PodCondition) error {
```
this function and  addConditionAndDeletePod could be combined into a single one and save some repetition and tests

mimowo replied:
Do you mean that instead of introducing the proxy function `markFailedAndDeletePod` we could just call `markFailedAndDeletePodWithCondition` with nil? Potentially, this was the initial implementation, but it was suggested in a previous review comment to introduce the proxy function to improve readability. I think I agree with this opinion, passing nil from multiple places often looks awkward. I don't think it would affect the set of testing scenarios.

krmayankk commented on test/integration/evictions/evictions_test.go:
```diff
@@ -334,6 +340,85 @@ func TestEvictionVersions(t *testing.T) {
+			enablePodDisruptionConditions: true,
+			phase:                         v1.PodRunning,
+		},
+		"running pod with job backoff policy disabled": {
```
nit: this gate is probably not `job backoff` but related to adding additional conditions about reason of pod eviction. In that context, this description is confusing

mimowo replied:
thanks, updated.

krmayankk commented on test/integration/evictions/evictions_test.go:
```diff
@@ -334,6 +340,85 @@ func TestEvictionVersions(t *testing.T) {
+			if tc.enablePodDisruptionConditions == true && cond == nil {
+				t.Errorf("Pod %q does not have the expected condition: %v", klog.KObj(updatedPod), v1.AlphaNoCompatGuaranteeDisruptionTarget)
+			} else if tc.enablePodDisruptionConditions == false && cond != nil {
+				t.Errorf("Pod %q does has the unexpected condition: %v", klog.KObj(updatedPod), v1.AlphaNoCompatGuaranteeDisruptionTarget)
```
nit: `Pod %q has an unexpected condition: %v`

mimowo replied:
Thanks, updated in a couple of places.

krmayankk commented on test/integration/podgc/podgc_test.go:
```diff
@@ -36,144 +37,224 @@ import (
+			if test.enablePodDisruptionConditions == true && cond == nil {
+				t.Errorf("Pod %q does not have the expected condition: %v", klog.KObj(pod), v1.AlphaNoCompatGuaranteeDisruptionTarget)
+			} else if test.enablePodDisruptionConditions == false && cond != nil {
+				t.Errorf("Pod %q does has the unexpected condition: %v", klog.KObj(pod), v1.AlphaNoCompatGuaranteeDisruptionTarget)
```
same nit as above

mimowo replied:
Updated in a couple of places

mimowo committed:
Introduction of a pod condition type indicating disruption. Its `reason` field indicates the reason:

- PreemptionByKubeScheduler (Pod preempted by kube-scheduler)
- DeletionByTaintManager (Pod deleted by taint manager due to NoExecute taint)
- EvictionByEvictionAPI (Pod evicted by Eviction API)
- DeletionByPodGC (an orphaned Pod deleted by PodGC)PreemptedByScheduler (Pod preempted by kube-scheduler)

mimowo commented:
/retest

aojea commented on pkg/controller/podgc/gc_controller_test.go:
```diff
@@ -136,29 +128,30 @@ func TestGCTerminated(t *testing.T) {
 					Spec:       v1.PodSpec{NodeName: "node"},
 				})
 			}
+			client := setupNewSimpleClient(nodes, pods)
+			gcc, podInformer, _ := NewFromClient(client, test.threshold)
+			for _, pod := range pods {
+				podInformer.Informer().GetStore().Add(pod)
+			}
```
this seems to differ from previous implementation, the client was only initialized with one node, now it seems to have pods too. 

However, I don't know how the clientset and the infomer initialized from the clientset works, is that the infomer automatically store the clientset initialized values?
if that is true, do we need to addd pods again to the informer storage?

mimowo replied:
Yes, they need to be added when constructing the client so that the patch operation does not fail with "not found" error. Actually, previously, the delete operation was failing with the NotFound error in the tests, but the tests didn't care - they only tested if the Delete action was received. Now, when the feature is enabled, I need the pods to be in the client in order to reach the delete operation. 

As for the need to add them to the pod informer - this is also needed (maybe it could be avoided with some refactoring), because if I don't add them to the informer then this list is empty: https://github.com/kubernetes/kubernetes/blob/719f3cf8da49d94da73b9ada1760e6f94e9834dd/pkg/controller/podgc/gc_controller.go#L111

aojea commented on pkg/controller/podgc/gc_controller_test.go:
```diff
@@ -317,11 +324,16 @@ func TestGCOrphaned(t *testing.T) {
+			for _, pod := range test.pods {
+				pods = append(pods, pod)
+			}
+			client := setupNewSimpleClient(nodes, pods)
```
I've done the same question here, is just a curiosity https://github.com/kubernetes/kubernetes/pull/110959/files#r935230821 , but if you are confident with the test behavior you can disregard it

mimowo replied:
See my response to the other place: https://github.com/kubernetes/kubernetes/pull/110959#discussion_r935309900

aojea commented on test/integration/evictions/evictions_test.go:
```diff
@@ -334,6 +340,85 @@ func TestEvictionVersions(t *testing.T) {
+				t.Fatalf("Failed to get the pod %q with error: %q", klog.KObj(pod), e)
+			}
+			_, cond := podutil.GetPodCondition(&updatedPod.Status, v1.PodConditionType(v1.AlphaNoCompatGuaranteeDisruptionTarget))
+			if tc.enablePodDisruptionConditions == true && cond == nil {
```
is there any possibilitiy the test can have a false negative with other condition or not the condition that we are exactly expecting?

mimowo replied:
I think it is reasonable to restrict the checks to the conditions which are involved in the recent code changes (DistruptionTarget in this case). Actually, most of the tests within evictions_test.go add another `Ready` condition, which I think we can skip in the assertions.

aojea commented:
> I left the hold since I couldn't tell if @aojea's comments on the tests were blocking

lgtm

soltysh reviewed (approved):
/lgtm
/hold cancel
since Antonio gave his :+1: above

alculquicondor commented:
/priority important-soon
